src/tests/test_blk_read_tracker.cpp

```diff
@@ -352,6 +358,198 @@ TEST_F(BlkReadTrackerTest, TestInsRmWithWaiterOverlapMultiReads2) {
+ * 2. do remove in massive threads (must be same amount of inserts);
+ * */
+TEST_F(BlkReadTrackerTest, TestThreadedInsertAndRemove) {
+    auto align = 8ul;
+    LOGINFO("Step 1: set entries per record to {}.", align);
+    get_inst()->set_entries_per_record(align);
+
+    std::vector< BlkId > bids{{10, 8, 0}, {20, 5, 0}, {25, 6, 0}, {43, 16, 0}, {56, 4, 0}, {72, 18, 0}, {122, 4, 0}};
+
+    const auto repeat = 100ul;
+    std::vector< std::thread > op_threads;
+
+    for (const auto& b : bids) {
+        std::thread t([this, &b]() {
+            for (auto j = 0ul; j < repeat; ++j) {
+                get_inst()->insert(b);
+            }
+        });
+        op_threads.push_back(std::move(t));
+    }
+
+    LOGINFO("Step 2: threaded insert issued.");
+
+    for (auto& t : op_threads) {
+        t.join();
+    }
+
+    // remove has to wait for insert to complete because if insert thread runs slower than remove, it might assert
+    // complaining no elements are found in map;
+
+    LOGINFO("Step 3: threaded insert joined.");
+    op_threads.clear();
+
+    for (const auto& b : bids) {
+        for (auto j = 0ul; j < repeat; ++j) {
+            std::thread t([this, &b]() { get_inst()->remove(b); });
+            op_threads.push_back(std::move(t));
+        }
+    }
+
+    LOGINFO("Step 4: threaded remove issued.");
+    for (auto& t : op_threads) {
+        t.join();
+    }
+
+    LOGINFO("Step 4: all threads joined.");
+}
+
+TEST_F(BlkReadTrackerTest, TestThreadedInsertWaitonThenRemove) {
+    auto align = 8ul;
+    LOGINFO("Step 1: set entries per record to {}.", align);
+    get_inst()->set_entries_per_record(align);
+
+    std::vector< BlkId > bids{{12, 6, 0}, {18, 5, 0}, {25, 8, 0}, {36, 16, 0}, {57, 4, 0}, {66, 18, 0}, {92, 14, 0}};
+    const auto repeat = 100ul;
+    std::vector< std::thread > op_threads;
+    for (const auto& b : bids) {
+        std::thread t([this, &b]() {
+            for (auto j = 0ul; j < repeat; ++j) {
+                get_inst()->insert(b);
+            }
+        });
+        op_threads.push_back(std::move(t));
+    }
+
+    LOGINFO("Step 2: threaded insert issued.");
+
+    std::vector< bool > called(bids.size(), false);
+    std::mutex mtx;
+    for (auto i = 0ul; i < bids.size(); ++i) {
+        std::thread t([this, &bids, i, &mtx, &called]() {
+            get_inst()->wait_on(bids[i], [i, &bids, &mtx, &called]() {
+                std::unique_lock lk(mtx);
+                LOGMSG_ASSERT(called[i] == false, "not expecting callback to be called more than once!");
+                called[i] = true;
+                LOGINFO("wait_on called on blkid: {};", bids[i].to_string());
+            });
+        });
+        op_threads.push_back(std::move(t));
+    }
+
+    // wait for all insert to complete, otherwise it will race with remove thread;
+    for (auto& t : op_threads) {
+        t.join();
+    }
+
+    op_threads.clear();
```

Comment:
Is it safe to clear op_threads as we are using it in line 445 for wait_on ? May be separate vec for wait on ?

Reply:
it won't clear until all threads are joined, right?

Reply:
Shouldnt remove call below cause the wait_on() callbacks to get triggered ? But if we clear here, is it still ok ?

Reply:
it is okay. The wait_on thread itself is just to simulate a free blk call, and once applied waiter, free operation thread was gone. The wait_on lambda is supposed to be called by remove operation's thread, which in real life is a read_completion routine's thread. I hope this explains.